### PR TITLE
QPPCT-762: Add better context to messages

### PR DESCRIPTION
**Pull request by halprin**
### Information
- JIRA story QPPCT-762.

### Changes proposed in this PR:
- Added another field to `Detail`, the class that gets serialized to report validation errors to our users, called `location`.
  - That field holds a human readable location.
- Gave human readable titles to certain `TemplateId`s.
- When an validation error occurs, the first `Node` or its parent that contains a human readable `TemplateId` is searched for.  The subsequently created `Detail` fills this human readable title as the `location`.
  - If this `Node` is a measure, the measure ID is obtained and appended to the human readable `location`.

### Checklist 
- [x] All JUnit tests pass (`mvn clean verify`).
- [x] New unit tests written to cover new functionality.
- [x] Added and updated JavaDocs for non-test classes and methods.
- [x] No local design debt. Do you feel that something is "ugly" after your changes?
- [n/a] Updated documentation (`README.md`, etc.) depending if the changes require it.

### CODE REVIEW

File: converter/src/main/java/gov/cms/qpp/conversion/model/TemplateId.java [hunk 1]
```diff
@@ -1,31 +1,32 @@
...
 import java.util.Map;
 import java.util.function.Function;
 
-import com.google.common.base.Strings;
-import gov.cms.qpp.conversion.util.EnvironmentHelper;
-
 /**
  * An enumeration of known templates IDs.
  */
 public enum TemplateId {
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/model/TemplateId.java]: What was your process for determining whether the template id has a human readable title?

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java [hunk 2]
```diff
@@ -1,31 +1,27 @@
...
+import java.util.Objects;
+
 /**
  * Holds the error information from Validators.
  */
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class Detail implements Serializable {
 	private static final long serialVersionUID = 8818544157552590676L;
 
-	@JsonProperty("errorCode")
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java]: glad we're cleaning this up

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java [hunk 3]
```diff
@@ -77,17 +76,35 @@ public static Detail forErrorCode(LocalizedError error) {
...
+		Node importantParentNode = node.findParentNodeWithHumanReadableTemplateId();
+
+		if (importantParentNode != null) {
+			String importantParentTitle = importantParentNode.getType().getHumanReadableTitle();
+			String possibleMeasureId = importantParentNode.getValue("measureId");
+
+			location = importantParentTitle;
+
+			if (!StringUtils.isEmpty(possibleMeasureId)) {
+				location += " " + possibleMeasureId;
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java]: do we want any extra separation here?

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java [hunk 4]
```diff
@@ -77,17 +76,35 @@ public static Detail forErrorCode(LocalizedError error) {
 		return detail;
 	}
 
+	private static String computeLocation(Node node) {
+
+		String location = null;
```
Comment (omnivion) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java]: use a stringbuilder

File: converter/src/main/java/gov/cms/qpp/conversion/model/TemplateId.java (hunk 1 above)
Comment (halprin) [reply converter/src/main/java/gov/cms/qpp/conversion/model/TemplateId.java]: The `TemplateId` that has a human readable title will have its title filled into the validation error location.

My thought was when it comes to a human wanting to know the location of an error, a measure will be the deepest that they care about.  If they want more information, they can use the line, column, and XPath information.

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java (hunk 3 above)
Comment (halprin) [reply converter/src/main/java/gov/cms/qpp/conversion/model/error/Detail.java]: What do you mean by separation?  As in extra lines between the code?

**Pull request merged**